Std/Util/Cache.lean

```diff
@@ -0,0 +1,160 @@
+/-
+Copyright (c) 2021 Gabriel Ebner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gabriel Ebner
+-/
+import Std.Lean.Meta.DiscrTree
+
+/-!
+# Once-per-file cache for tactics
+
+This file defines cache data structures for tactics
+that are initialized the first time they are accessed.
+Since Lean 4 starts one process per file, these caches are once-per-file
+and can for example be used to cache information about the imported modules.
+
+The `Cache α` data structure is the most generic version we define.
+It is created using `Cache.mk f` where `f : MetaM α` performs the initialization of the cache:
+```
+initialize numberOfImports : Cache Nat ← Cache.mk do
+  (← getEnv).imports.size
+
+-- (does not work in the same module where the cache is defined)
+#eval show MetaM Nat from numberOfImports.get
+```
+
+The `DeclCache α` data structure computes a fold over the environment's constants:
+`DeclCache.mk empty f` constructs such a cache
+where `empty : α` and `f : Name → ConstantInfo → α → MetaM α`.
+The result of the constants in the imports is cached between tactic invocations,
+while for constants defined in the same file `f` is evaluated again every time.
+This kind of cache can be used e.g. to populate discrimination trees.
+-/
+
+open Lean Meta
+
+namespace Mathlib.Tactic
```

Comment:
Did you intentionally keep using the `Mathlib` namespace even in std, @semorrison ?

Reply:
Uh, no, sorry, this was an oversight.